// gupshup-utils/lib/StructuredMessages.js
/**
 * Created by agnibha on 2/5/17.
 */

module.exports.Poll = class Poll {
    constructor(question, msgId) {
        this.type = "poll";
        this.question = question;
        this.msgId = msgId || "";
    }

    getQuestion() {
        return this.question;
    }

    setQuestion(question) {
        this.question = question;
    }

    setMsgId(msgId) {
        this.msgId = msgId;
    }

    getMsgId() {
        return this.msgId;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};

module.exports.Survey = class Survey {
    constructor(question, msgId, options) {
        this.type = "survey";
        this.question = question;
        this.msgId = msgId || "";
        this.options = options || [];
    }

    getQuestion() {
        return this.question;
    }

    setQuestion(question) {
        this.question = question;
    }

    getMsgId() {
        return this.msgId;
    }

    setMsgId(msgId) {
        this.msgId = msgId;
    }

    getOptions() {
        return this.options;
    }

    setOptions(options) {
        this.options = options;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};

module.exports.Catalogue = class Catalogue {
    constructor(items, imageaspectratio, msgId) {
        this.type = 'catalogue';
        this.items = items;
        this.imageaspectratio = imageaspectratio || 'horizontal';
        this.msgId = msgId || '';
    }

    getImageAspectRatio() {
        return this.imageaspectratio;
    }

    setImageAspectRatio(imageaspectratio) {
        this.imageaspectratio = imageaspectratio;
    }

    getMsgId() {
        return this.msgId;
    }

    setMsgId(msgId) {
        this.msgId = msgId;
    }

    getItems() {
        return this.items;
    }

    setItems(items) {
        this.items = items;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};

module.exports.CatalogueItem = class CatalogueItem {
    constructor(title, subtitle, imgUrl, options) {
        this.title = title || "";
        this.subtitle = subtitle || "";
        this.imgUrl = imgUrl || "";
        this.options = options || [];
    }

    getTitle() {
        return this.title;
    }

    setTitle(title) {
        this.title = title;
    }

    getSubtitle() {
        return this.subtitle;
    }

    setSubtitle(subtitle) {
        this.subtitle = subtitle;
    }

    getImgUrl() {
        return this.imgUrl;
    }

    setImgUrl(imgUrl) {
        this.imgUrl = imgUrl;
    }

    getOptions() {
        return this.options;
    }

    setOptions(options) {
        this.options = options;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};

module.exports.Option = class Option {
    constructor(type, title, url, iconurl) {
        this.type = type;
        this.title = title;
        this.url = url;
        this.iconurl = iconurl;
    }

    getType() {
        return this.type;
    }

    setType(type) {
        this.type = type;
    }

    getTitle() {
        return this.title;
    }

    setTitle(title) {
        this.title = title;
    }

    getUrl() {
        return this.url;
    }

    setUrl(url) {
        this.url = url;
    }

    getIconUrl() {
        return this.iconurl;
    }

    setIconUrl(iconurl) {
        this.iconurl = iconurl;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};

module.exports.QuickReply = class QuickReply {
    constructor(content, msgid, options) {
        this.type = 'quick_reply';
        this.content = content;
        this.msgid = msgid;
        this.options = options;
    }

    getContent() {
        return this.content;
    }

    setContent(content) {
        this.content = content;
    }

    getMsgId() {
        return this.msgId;
    }

    setMsgId(msgid) {
        this.msgid = msgid;
    }

    getOptions() {
        return this.options;
    }

    setOptions(options) {
        this.options = options;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
};
