// gupshup-utils/lib/URLShortener.js
/**
 * Created by agnibha on 15/5/17.
 */
const request = require("request");
const memCache = {};
const has = Object.prototype.hasOwnProperty;
module.exports = class URLShortener {
    getShortenedURL(url, callback) {
        if (has.call(memCache, url)) {
            callback(memCache[url]);
        } else {
            makeShortURL(url, (shortenedUrl) => {
                memCache[url] = shortenedUrl;
                callback(shortenedUrl);
            });
        }
    }
};

function makeShortURL(url, callback) {
    "use strict";
    let body = {
        "key": "AIzaSyD4PH1D8WF26E-nt2QWr6ZnMLJQffouDkM",
        "longUrl": url
    };
    let options = {
        method: 'POST',
        url: 'http://gs.tc.im/urlshortener/v1/url',
        headers: {
            'content-type': 'application/json',
            'cache-control': 'no-cache'
        },
        body: JSON.stringify(body)
    };

    request(options, function (error, response, body) {
        if (error) throw new Error(error);
        callback(JSON.parse(body).shortURL);
    });
}
